rfc/curio-api-rfc.md

```diff
@@ -0,0 +1,157 @@
+# Curio API RFC
+
+This is a proposed set of APIs to be implemented in Curio to interface with a storacha node.
+
+The proposed flow has three core API endpoints:
+1. Endpoints for manipulating proofs sets
+2. An endpoint for storing pieces + flow for storing pieces
+3. An endpoint for retrieving pieces
+
+There are addition considerations we should consider:
+1. Authorization -- In storacha's network, it's important that the original end user maintain control of authorization for any action performed (including retrieval). We accomplish this through UCANs. We should discuss how we can maintain this without forcing curio to implement a full UCAN authorization process.
+2. Aggregation - storacha's data is at times extremely small (<1mb in certain cases). Our understanding is that economically, it makes more sense to do some light aggregation of data before adding it to the proof set. The proposal below outlines a facility for doing this. While storacha would store pieces as it receives them, we would add them to the proof set in a seperate step, with a root that could optionally be an aggregate of several pieces.
+3. IPNI announcements -- we plan to use IPNI announcements in a specific way with our pieces. Our understanding is the curio IPNI flow is in flux. We can try to integrate your IPNI api or just do it ourselves.
```

Comment:
IPNI is pretty much implemented now, with the following schema that coordinates it on the curio side:

```
-- Table for storing IPNI ads
CREATE TABLE ipni (
    order_number BIGSERIAL PRIMARY KEY, -- Unique increasing order number
    ad_cid TEXT NOT NULL,
    context_id BYTEA NOT NULL, -- abi.PieceInfo in Curio
    -- metadata column in not required as Curio only supports one type of metadata(HTTP)
    is_rm BOOLEAN NOT NULL,

    previous TEXT, -- previous ad will only be null for first ad in chain

    provider TEXT NOT NULL, -- peerID from libp2p, this is main identifier on IPNI side
    addresses TEXT NOT NULL, -- HTTP retrieval server addresses

    signature BYTEA NOT NULL,
    entries TEXT NOT NULL, -- CID of first link in entry chain

    unique (ad_cid)
);

CREATE TABLE ipni_head (
    provider TEXT NOT NULL PRIMARY KEY, -- PeerID from libp2p, this is the main identifier
    head TEXT NOT NULL, -- ad_cid from the ipni table, representing the head of the ad chain

    FOREIGN KEY (head) REFERENCES ipni(ad_cid) ON DELETE RESTRICT -- Prevents deletion if it's referenced
);

-- This table stores metadata for ipni ad entry chunks. This metadata is used to reconstruct the original ad entry from
-- on-disk .car block headers or from data in the piece index database.
CREATE TABLE ipni_chunks (
    cid TEXT PRIMARY KEY, -- CID of the chunk
    piece_cid TEXT NOT NULL, -- Related Piece CID
    chunk_num INTEGER NOT NULL, -- Chunk number within the piece. Chunk 0 has no "next" link.
    first_cid TEXT, -- In case of db-based chunks, the CID of the first cid in the chunk
    start_offset BIGINT, -- In case of .car-based chunks, the offset in the .car file where the chunk starts
    num_blocks BIGINT NOT NULL, -- Number of blocks in the chunk
    from_car BOOLEAN NOT NULL, -- Whether the chunk is from a .car file or from the database
    CHECK (
        (from_car = FALSE AND first_cid IS NOT NULL AND start_offset IS NULL) OR
        (from_car = TRUE AND first_cid IS NULL AND start_offset IS NOT NULL)
    ),

    UNIQUE (piece_cid, chunk_num)
);
```

Now, IPNI likes larger ads, so ideall storacha would create aggregate ads for multiple pieces; we can extend `ipni_chunk` to support reading from storacha-stored pieces (though really technically just the piececid works fine there)